counsel.el

```diff
@@ -2617,18 +2617,27 @@ string - the full shell command to run."
   "Use `dired-jump' on X."
   (dired-jump nil x))
 
+(defun counsel--locate-executable ()
+  "Return an available `locate' implementation's executable name."
+  (cond
+   ((counsel-require-program "plocate" 'noerror)
+    "plocate")
+   ((counsel-require-program "locate")
+    "locate")))
```

Comment:
These checks should be user-configurable, and can happen at load-time instead of run-time, with `executable-find`.

So I suggest either extending `counsel-locate-cmd` accordingly, with new function entries for `plocate` as appropriate (preferred), or adding a new user option altogether (if necessary).

Reply:
> `counsel-locate-cmd` accordingly, with new function entries

I'm afraid the that would lead to code duplication, since we already have: `counsel-locate-cmd-noregex` and `counsel-locate-cmd-default`,  both have hard-coded `"locate"`; if we opt to adding more options to `counsel-locate-cmd`, wouldn't it produce 2 more functions: `counsel-plocate-cmd-default`, `counsel-plocate-cmd-noregex`, and all 4 of them are similar?

Also, the reason I chose to define a "dynamic" `#'counsel--locate-executable` was `locate` and `plocate` were mutually exclusive, it's impossible to install more than one of them under most package managers, therefore we should simply utilize the available one since it's the only option available (maybe I should make a configurable list like ` defcustom counsel-locate-executable-list '("locate" "plocate")` from the start and check each one of them, for extensibility in case a new implementation pops up in the future).

> a new user option altogether (if necessary).

I think honoring built-in `locate-command` defined in `locate.el` should be the simplest route (although I myself prefer zero-configuration as a user).

I am looking forward for your choice.

Edit: clarification, grammar.

Reply:
> `counsel-locate-cmd` accordingly, with new function entries

I'm afraid the that would lead to code duplication, since we already have: `counsel-locate-cmd-noregex` and `counsel-locate-cmd-default`,  both have hard-coded `"locate"`; if we opt to adding more options to `counsel-locate-cmd`, wouldn't it produce 2 more functions: `counsel-plocate-cmd-default`, `counsel-plocate-cmd-noregex`, and all 4 of them are similar?

Also, the reason I chose to define a "dynamic" `#'counsel--locate-executable` was `locate` and `plocate` were mutually exclusive, it's impossible to install more than one of them under most package managers, therefore we should simply utilize the available one since it's the only option available (maybe I should make a configurable list like ` defcustom counsel-locate-executable-list '("locate" "plocate")` from the start and check each one of them, for extensibility in case a new implementation pops up in the future).

> a new user option altogether (if necessary).

I think honoring built-in `locate-command` defined in `locate.el` should be the simplest route (although I myself prefer zero-configuration as a user).

I am looking forward for your choice.

Reply:
> and all 4 of them are similar?

Yes, I never said I was a fan of the current "design" ;).

> `locate` and `plocate` were mutually exclusive

It's not our job to try to outsmart the user or their system. It is important for the user to be able to control from within Emacs which external executables are invoked, and by what name.

> I think honoring built-in `locate-command` defined in locate.el should be the simplest route

SGTM. See also #385.

> (although I myself prefer zero-configuration as a user)

FWIW, `update-alternatives` on Debian indirectly symlinks `/usr/bin/locate` to `/usr/bin/plocate` by default, i.e. installing `plocate` makes it the default `locate` on my system.

Reply:
> FWIW, `update-alternatives` on Debian indirectly symlinks `/usr/bin/locate` to `/usr/bin/plocate` by default, i.e. installing `plocate` makes it the default `locate` on my system.

I didn't know it, if only `plocate` did that by default.

> > I think honoring built-in `locate-command` defined in locate.el should be the simplest route
> 
> SGTM. See also #385.

I have pushed #2859 as a replacement, please close this PR as you judge which is more appropriate.

Reply:
A side note: I have completed Copyright Assignment, do I have to provide the paper?

Reply:
> A side question: I have completed Copyright Assignment

Welcome aboard! :)

> do I have to provide the paper?

No, I just need to ping a GNU maintainer with access to the CA list to check that you're on it.  Once they've confirmed (Eli, one of the Emacs maintainers, usually responds within a day), I can look into merging your PRs :).

[There was talk a while ago about setting this information up as a decentralised service for the benefit of people like me who don't have access to a GNU shell account, but I haven't heard about it since.]